### ISSUE TITLE
Crawler stopped importing data

### ISSUE THREAD

**Issue by leonid-deriv** [change-metrics/monocle#1112]
I have noticed that a crawler stopped importing data. I see the following errors in the log 

```
2024-01-31 14:30:00 INFO    Macroscope.Worker:183: Looking for oldest entity {"index":"demo","crawler":"xxxx-monocle-demo","stream":"TaskDatas","offset":0}
2024-01-31 14:30:00 INFO    Macroscope.Worker:199: Processing {"index":"demo","crawler":"xxxx-monocle-demo","stream":"Changes","entity":{"contents":"xxxx/yyyy","tag":"Project"},"age":"2023-11-29T01:41:16Z"}
2024-01-31 14:30:00 WARNING Lentille.GitHub.RateLimit:66: Repository not found. Will not retry. {"index":"demo","crawler":"xxxx-monocle-demo","stream":"Changes"}
2024-01-31 14:30:00 INFO    Lentille.GraphQL:232: Fetched from current page {"index":"demo","crawler":"xxxx-monocle-demo","stream":"Changes","count":0,"total":0,"pageInfo":{"endCursor":null,"hasNextPage":false,"totalCount":null},"ratelimit":null}
2024-01-31 14:30:00 WARNING Lentille.GraphQL:276: Fetched partial result {"index":"demo","crawler":"xxxx-monocle-demo","stream":"Changes","err":[{"locations":[{"column":7,"line":8}],"message":"Could not resolve to a Repository with the name 'xxxx/yyyy'.","path":["repository"],"type":"NOT_FOUND"}]}
2024-01-31 14:30:00 INFO    Macroscope.Worker:204: Posting documents {"index":"demo","crawler":"xxxx-monocle-demo","stream":"Changes","count":2}
2024-01-31 14:30:00 INFO    Macroscope.Worker:189: Unable to find entity to update {"index":"demo","crawler":"xxxx-monocle-demo","stream":"TaskDatas"}
2024-01-31 14:30:00 INFO    Macroscope.Worker:183: Looking for oldest entity {"index":"demo","crawler":"xxxx-monocle-demo","stream":"Changes","offset":0}
2024-01-31 14:30:00 WARNING Macroscope.Worker:167: Stream produced a fatal error {"index":"demo","crawler":"xxxx-monocle-demo","stream":"Changes","err":["2024-01-31T14:30:00.901152636Z",{"contents":["Unknown GetProjectPullRequests response: GetProjectPullRequests {rateLimit = Just (GetProjectPullRequestsRateLimit {used = 183, remaining = 4817, resetAt = DateTime \"2024-01-31T14:53:42Z\"}), repository = Nothing}"],"tag":"DecodeError"}]}
```

Actually, the repository which cannot be found does not exist. 
I thought it could be cached so I have restarted services but looks like still have the problem
any suggestions?

**Comment by leonid-deriv** [comment]
Maybe this is a coincidence, but the data stopped being imported on the date when I ran the group update which you know failed.

**Comment by leonid-deriv** [comment]
One more comment. xxxx/yyyy - not sure why crawler is trying to get this repo. I did execute both REST and GraphQL requests and this repo is not returned by GitHub. This repo is not shown in Monocle WEB interface.

**Comment by leonid-deriv** [comment]
Had to reindex :(

**Comment by leonid-deriv** [comment]
Looks like we have a similar case again. After some "event" it stops importing data :(.  symptoms similar to what described before.   What I remember that the repository the crawler complains about did not exist ... this time I also cannot find this report. Last time the only solution was to completely rebuild the index but I am afraid this is not a good option.  Any idea how we can troubleshoot it. Here is another error message I see regularly in the log 

```
2024-03-07 19:16:09 WARNING Macroscope.Worker:167: Stream produced a fatal error {"index":"xxxx","crawler":"xxx-monocle-xxxx","stream":"Changes","err":["2024-03-07T19:16:09.507202765Z",{"contents":["Unknown GetProjectPullRequests response: GetProjectPullRequests {rateLimit = Just (GetProjectPullRequestsRateLimit {used = 130, remaining = 4870, resetAt = DateTime \"2024-03-07T19:32:37Z\"}), repository = Nothing}"],"tag":"DecodeError"}]}
```

**Comment by leonid-deriv** [comment]
To me, taking into account that it refers to non existing repo, some internal cache? maybe corrupted. So maybe it is possible to clean it and then I can reset the date to re-scan data? I really  do not want to re-index it again, plus now it happened for the second time so probably will happen again :(

**Comment by leonid-deriv** [comment]
another question about  last date.  `Monocle crawlers keep track of the last date (commit date) when a successful document fetch happened.` 
Where crawler stores this data.

**Comment by morucci** [comment]
Yes there is a cache. The CLI does not provide a way to clear such entries for no longer existing repositories.
Perhaps then you could try to remove the related state object in the Elasticsearch DB `https://github.com/change-metrics/monocle/blob/master/src/Monocle/Backend/Index.hs#L204`

**Comment by leonid-deriv** [comment]
I am trying to find the index in the elastic where you store this metadata
and cannot find it. it is not visible in Kibana - or I am doing something
wrong

**Comment by morucci** [comment]
I think because that's an object without the usual date field. So you need to select the right parameter in the kibana index pattern creation.

**Comment by leonid-deriv** [comment]
but what is the index name? it is not the index where all workspace data is
stored?

**Comment by morucci** [comment]
same index

**Comment by leonid-deriv** [comment]
Fabien, sorry for the trouble. But I cannot find the required information.
What is the document type to cache crawler information?

**Comment by leonid-deriv** [comment]
Was not very attentive looking at the index. removing a repo from Elastic
looks like solved the problem.
Should I register a bug for it?

Leonid

**Comment by morucci** [comment]
Hi, thanks you to have confirmed this. We can just keep that issue for us to investigate the fact that the crawler stop when a no longer existing is still in the "cache" Such objects can stay in the cache but should not prevent the crawler to process the rest of the repo.

**Comment by TristanCacqueray** [comment]
Looking at it, it looks like:

- the error comes from (when the repositorty is empty): https://github.com/change-metrics/monocle/blob/d2232bffee1a381c991854dea7abc6784926137c/src/Lentille/GitHub/PullRequests.hs#L83 
- and this stops the PR crawler because the postResult considered this is a fatal error: https://github.com/change-metrics/monocle/blob/d2232bffee1a381c991854dea7abc6784926137c/src/Macroscope/Worker.hs#L208-L217

I think we could:

- add a new `EntityRemoved` error to https://github.com/change-metrics/monocle/blob/d2232bffee1a381c991854dea7abc6784926137c/src/Lentille.hs#L113-L114 
- cleanup the crawler metadata when it happens in the Worker module
- ignore it to keep the crawler running in https://github.com/change-metrics/monocle/blob/d2232bffee1a381c991854dea7abc6784926137c/src/Macroscope/Worker.hs#L113-L119

> Note that the comment above is not correct, it should says `This is likely an error we *can* recover`

**Comment by leonid-deriv** [comment]
thank you, the most important is to make this error "non-fatal" so a
crawler continues running. And all your 3 points make sense.

**Comment by leonid-deriv** [comment]
Any chance to fix this error? The problem is that we are dropping "old" repos and I have to manually remove it from cache every single time :(

**Comment by gekitsuu** [comment]
What's the easiest way to clear the cache?

**Comment by christophe-kamphaus-jemmic** [comment]
I also ran into this issue.

> What's the easiest way to clear the cache?

For me using [elasticvue](https://github.com/cars10/elasticvue) to connect to Elastic allowed me to search for any documents containing the deleted repo (ie the cached repo info as well as all crawler errors for this repo) and mass deleting them. This allowed the crawler to resume crawling all my repos.

**Comment by leonid-deriv** [comment]
I connect to Elastic using Kibana, search for repo - for example 
  - crawler_metadata.crawler_type_value:"owner/repo" 
  - get the _id of the doc
  -  delete the doc (for example from Kibana devtools)

**Comment by morucci** [comment]
Please try with version 1.11.2, that we just released.

**Comment by gekitsuu** [comment]
@morucci it's working for me now, so I _think_ this addressed my issue with repos that had been deleted. I have a co-worker facing the same problem so I'll have her test it too and report back.

One confusing thing I did notice is that it looks like the UI is reporting 1.11.1 in the 1.11.2 image which confused me a bit until I confirmed in Docker that I was running the right label.

<img width="514" alt="Screenshot 2024-12-24 at 12 24 14 PM" src="https://github.com/user-attachments/assets/4928f675-1df7-46b5-84aa-e477ad765e64" />
<img width="890" alt="Screenshot 2024-12-24 at 12 24 19 PM" src="https://github.com/user-attachments/assets/1768fd01-44c3-4d7c-a291-05a2c2b1c132" />

**Comment by morucci** [comment]
Thanks for the feedback. I need investigate why we get the wrong version number.

**Comment by morucci** [comment]
@gekitsuu Yes you are right and the wrong version displaying is due to my mistake during the release process. Referring to the commit hash confirms that you tested the right version https://github.com/change-metrics/monocle/commit/bfe857fe580476e20072ccb7275b026f3481253c. The Monocle API returns the version number based on the Cabal file `https://github.com/change-metrics/monocle/blob/master/monocle.cabal#L3`  that I missed to update prior to the Git tag, that trigger the our release CI workflow.

I'll close this issue as you confirmed the issue is fixed. Please re-open if needed.
